windows/hid.c

```diff
@@ -1048,6 +1048,44 @@ int HID_API_EXPORT_CALL HID_API_CALL hid_get_indexed_string(hid_device *dev, int
 	return 0;
 }
 
+int HID_API_EXPORT_CALL hid_winapi_get_container_id(hid_device *dev, GUID *guid)
+{
+	GUID *container_id = NULL;
+	wchar_t *interface_path = NULL, *device_id = NULL;
+	CONFIGRET cr;
+	DEVINST dev_node;
+
+	if (!guid)
+		return -1;
+
+	interface_path = hid_internal_UTF8toUTF16(dev->device_info->path);
+	if (!interface_path)
+		goto end;
+
+	/* Get the device id from interface path */
+	device_id = hid_internal_get_device_interface_property(interface_path, &DEVPKEY_Device_InstanceId, DEVPROP_TYPE_STRING);
+	if (!device_id)
+		goto end;
+
+	/* Open devnode from device id */
+	cr = CM_Locate_DevNodeW(&dev_node, (DEVINSTID_W)device_id, CM_LOCATE_DEVNODE_NORMAL);
+	if (cr != CR_SUCCESS)
+		goto end;
+
+	/* Get the container id from devnode */
+	container_id = hid_internal_get_devnode_property(dev_node, &DEVPKEY_Device_ContainerId, DEVPROP_TYPE_GUID);
+	if (container_id) {
+		*guid = *container_id;
+	}
```

Comment:
call `CM_Get_DevNode_PropertyW` directly and pass `guid` - you'll avoid extra memory allocation/copy/free